File: problems/13_roman_to_Integer/roman_to_integer_JAVA/roman_to_integer.java
package roman_to_integer_JAVA;
import java.util.Map;
import java.util.HashMap;

/**
 * Solution for LeetCode Problem 13: Roman to Integer.
 *
 * This class provides a method to convert a Roman numeral string into its integer value.
 * The conversion works from right to left, subtracting values when a smaller numeral precedes a larger one,
 * and adding otherwise, according to Roman numeral rules.
 */
class Solution {
    /**
     * Converts a Roman numeral string to its integer representation.
     *
     * @param s the Roman numeral string (e.g., "MCMXCIV")
     * @return the integer value of the Roman numeral
     */
    public int romanToInt(String s) {
        // Map each Roman numeral to its integer value
        Map<Character, Integer> romanMap = new HashMap<>();
        romanMap.put('I', 1);
        romanMap.put('V', 5);
        romanMap.put('X', 10);
        romanMap.put('L', 50);
        romanMap.put('C', 100);
        romanMap.put('D', 500);
        romanMap.put('M', 1000); 

        int result = 0;
        int prevValue = 1; // Previous numeral value (start with 1 for the rightmost character)

        // Iterate from right to left through the string
        for (int i = s.length() - 1; i >= 0; i--) {
            int currValue = romanMap.get(s.charAt(i));
            if (currValue < prevValue) {
                // If current value is less than previous, subtract it
                result -= currValue; 
            } else {
                // Otherwise, add it
                result += currValue; 
            }
            prevValue = currValue; 
        } 
        return result;
    } 
}

File: problems/13_roman_to_Integer/roman_to_integer_TS/roman_to_integer.ts
/**
 * LeetCode Problem 13: Roman to Integer
 * 
 * Converts a Roman numeral string to its integer representation.
 * Handles standard Roman numeral notation, including subtractive combinations
 * like IV (4), IX (9), XL (40), XC (90), CD (400), and CM (900).
 *
 * @param {string} s - A string containing the Roman numeral (e.g., "MCMXCIV").
 * @return {number} - The integer value corresponding to the Roman numeral.
 *
 * @example
 * console.log(romanToInt("III"));      // Output: 3
 * console.log(romanToInt("LVIII"));    // Output: 58
 * console.log(romanToInt("MCMXCIV"));  // Output: 1994
 */
function romanToInt(s: string): number {
    // Mapping of Roman numerals to integer values
    const romanMap: { [key: string]: number } = {
        'I': 1,
        'V': 5,
        'X': 10,
        'L': 50,
        'C': 100,
        'D': 500,
        'M': 1000
    };

    let result = 0;

    // Iterate through the string, subtracting when a smaller numeral precedes a larger one
    for (let i = 0; i < s.length - 1; i++) {
        if (romanMap[s[i]] < romanMap[s[i + 1]]) {
            result -= romanMap[s[i]];
        } else {
            result += romanMap[s[i]];
        }
    }

    // Add the value of the last numeral
    return result + romanMap[s[s.length - 1]];
}

// Example execution for CI test
console.log(romanToInt("MCMXCIV")); // Output: 1994